migrate-to-w3up.js

```diff
@@ -0,0 +1,173 @@
+const isMain = (url, argv = process.argv) => fileURLToPath(url) === fs.realpathSync(argv[1])
+if (isMain(import.meta.url, process.argv)) {
+  main(process.argv).catch(error => console.error('error in main()', error))
+}
+
+/**
+ * get w3up-client with store from a good default store
+ */
+async function getDefaultW3up() {
+  // instead, accept accept W3_PRINCIPAL and W3_PROOF env vars or flags 
+  const store = new StoreConf({ profile: process.env.W3_STORE_NAME ?? 'w3cli' })
+  const w3 = await w3up.create({ store })
+  return w3
+}
+
+/**
+ * get a @web3-storage/access/agent instance with default store
+ */
+async function getDefaultW3upAgent() {
+  const w3 = await getDefaultW3up()
+  // @ts-expect-error _agent is protected property
+  const access = w3._agent
+  return access
+}
+
+/**
+ * main function that runs when this file is executed.
+ * It reads configuration from cli args, env vars, etc,
+ * then runs a migration using the `migration` function defined above.
+ * @param {string[]} argv - command line arguments
+ */
+async function main(argv) {
+  const { values } = parseArgs({
+    options: {
+      space: {
+        type: 'string',
+        help: 'space DID to migrate to',
+      }
+    }
+  })
+
+  const agent = await getDefaultW3upAgent()
+  // source of uploads is stdin by default
+  /** @type {AsyncIterable<W32023Upload>} */
+  let source
+  let isInteractive
+  // except stdin won't work if nothing is piped in.
+  // If nothing piped in, ask the user what to do.
+  if (!process.stdin.isTTY) {
+    source = new W32023UploadsFromNdjson(Readable.toWeb(process.stdin))
+  } else {
+    source = await getUploadsFromPrompts()
+    isInteractive = true
+  }
+
+  let spaceValue = values.space
+  if (isInteractive && !spaceValue) {
+    const chosenSpace = await promptForSpace()
+    console.warn('using space', chosenSpace.did())
+    spaceValue = chosenSpace.did()
+  }
+  if ( ! spaceValue) {
+    throw new Error(`pass migration destination with \`--space <space.did>\`.`)
+  }
+  const space = DID.match({ method: 'key' }).from(spaceValue)
+
+  const migration = migrate({
+    issuer: agent.issuer,
+    w3up: agent.connection,
+    source: Readable.toWeb(Readable.from(source)),
+    destination: new URL(space),
+    async fetchPart(cid, { signal }) {
+      return await fetch(new URL(`/ipfs/${cid}`, 'https://w3s.link'), { signal })
+    },
+    authorization: agent.proofs([
+      {
+        can: 'store/add',
+        with: space,
+      },
+      {
+        can: 'upload/add',
+        with: space,
+      },
+    ])
+  })
+  for await (const event of migration) {
+    console.log(JSON.stringify(event, stringifyForMigrationProgressStdio, isInteractive ? 2 : undefined))
+  }
+}
+
+/**
+ * JSON.stringify replacer for progress of migration
+ * @param {string} key - json property name
+ * @param {any} value - json property value
+ */
+function stringifyForMigrationProgressStdio(key, value) {
+  if (key === 'receipt') {
+    return {
```

Comment:
@Gozala This could be replaced by https://github.com/web3-storage/ucanto/pull/339/files